components/sup/src/cli_test_helpers.rs

```diff
@@ -0,0 +1,151 @@
+use clap::{App, ArgMatches};
+use std::fmt;
+use std::iter::FromIterator;
+
+#[derive(Debug)]
+pub enum Expectation<'a> {
+    Value(&'a str),
+    Values(Vec<String>),
+}
+
+use self::Expectation::{Value, Values};
+
+#[derive(Debug)]
+enum CliTestError {
+    ValueMismatch(String, String, String),
+    MultiValueMismatch(String, Vec<String>, Vec<String>),
+    MissingFlag(String),
+    MissingSubcmd(),
+    ClapError(String),
+}
+
+impl fmt::Display for CliTestError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let content = match self {
+            CliTestError::ValueMismatch(flag, expected, actual) => format!(
+                r#"
+'{}' validation failed:
+  Expected: '{}'
+  Actual:   '{}'
+"#,
+                flag, expected, actual
+            ),
+            CliTestError::MultiValueMismatch(flag, expected, actual) => format!(
+                r#"
+'{}' validation failed:
+  Expected: '{:#?}'
+  Actual:   '{:#?}'
+"#,
+                flag, expected, actual,
+            ),
+            CliTestError::MissingFlag(flag) => format!("Expected {} flag was missing.", flag),
+            CliTestError::MissingSubcmd() => {
+                format!("Expected a subcommand, but none was specified")
+            }
+            CliTestError::ClapError(err) => format!(
+                r#"
+Command Parse error:
+****************************************
+{}
+****************************************
+"#,
+                err.to_string()
+            ),
+        };
+        write!(f, "{}", content)
+    }
+}
+
+pub fn assert_command(app: App, cmd: &str, assertions: Vec<(&str, Expectation)>) {
+    let cmd_vec = Vec::from_iter(cmd.split_whitespace());
+    let errors: Vec<CliTestError> = match app.get_matches_from_safe(cmd_vec) {
+        Ok(matches) => match matches.subcommand() {
+            (_, Some(matches)) => assert_matches(matches, assertions),
+            (_, None) => vec![CliTestError::MissingSubcmd()],
+        },
+        Err(err) => vec![CliTestError::ClapError(err.to_string())],
+    };
+
+    if !errors.is_empty() {
+        let error_string = errors
+            .into_iter()
+            .map(|error| format!("{}", error))
+            .collect::<Vec<String>>()
+            .join("\n");
+
+        panic!(
+            r#"
+
+Failed assertions for command: '{}'
+
+{}
+
+"#,
+            cmd, error_string
+        );
+    }
+}
+
+fn assert_matches(matches: &ArgMatches, assertions: Vec<(&str, Expectation)>) -> Vec<CliTestError> {
+    assertions
+        .into_iter()
+        .fold(Vec::new(), |mut errs, (flag, expected_value)| {
```

Comment:
I'm a big fan of FP, but in this case I don't think it's adding to the legibility of this code. I'd try writing it out with a `for` loop over the assertions and see if that ends up shorter/simpler.

Mostly I'm not a fan of using `fold` when the computation in the closure doesn't really depend on the memoized object*. It's possible I would like this better if it were a `map` of assertion → error. But since assertions may map to no error, I think a [`filter_map`](http://doc.rust-lang.org/1.29.2/core/iter/trait.Iterator.html#method.filter_map) would be appropriate and your closure would return `Option<CliTestError>` and a `collect` would leave you with `Vec<CliTestError>`.

\* I know `errs` carries the state, but that's not really the same thing in my mind.